proposals/0000-unsatisfiable.rst

```diff
@@ -0,0 +1,599 @@
+  e`` remains unsolved, a warning diagnostic is emitted, displaying the message
+  that results from normalising and rendering the type ``e``.
+
+* An ``Warning`` constraint is never automatically generalised.
+
+* The presence of ``Warning`` constraints does not affect pattern-match coverage
+  checking or the functional dependency check.
+
+* GHC will report an error if a user attempts to define an instance for
+  ``Warning``.
+
+* The representation of an ``Warning e`` constraint in Core is equivalent to the
+  dictionary ``data WarningDict e = WarningDict``.  This is GHC's normal
+  representation of a class with no methods.
+
+See `#17027 <https://gitlab.haskell.org/ghc/ghc/-/issues/17027>`_ for a request
+for a feature like this.  Unlike the suggestion in that ticket, the ``Warning``
+class proposed here can only be used at kind ``Constraint``, for similar reasons
+to ``Unsatisfiable``.
+
+The warning diagnostics emitted by this feature are controlled by a new warning
+class ``-Wtype-warnings``, which is enabled by default, but may be disabled with
+``-Wno-type-warnings``.
+
+This provides a more flexible alternative to ``WARNING`` and ``DEPRECATED``
+pragmas, because warnings may be displayed only when definitions are used at
+particular types, and the messages may be constructed dynamically.
+
+(It would be possible to add a ``Deprecated`` constraint, similar to
+``Warning``, specifically for use with deprecation warnings, and controlled by a
+different flag.  However separating out deprecation warnings specifically does
+not seem strongly motivated.)
+
+
+
+Examples
+--------
+
+For more substantial examples, see the `unsatisfiable package by Oleg Grenrus
+<https://hackage.haskell.org/package/unsatisfiable>`_.  This package implements
+a type-checker plugin that roughly corresponds to the design of the
+``Unsatisfiable`` constraints in this proposal.
+
+#. The following definitions are accepted (assuming ``DataKinds``,
+   ``FlexibleContexts`` and ``TypeApplications`` are enabled)::
+
+     type Msg = Text "Cannot call 'uncallable'."
+
+     uncallable :: Unsatisfiable Msg => ()
+     uncallable = unsatisfiable @Msg
+
+     uncallable' :: Unsatisfiable Msg => ()
+     uncallable' = uncallable
+
+   The definition of ``uncallable`` is accepted because ``Unsatisfiable`` appears
+   as a Given, not a Wanted.  In ``uncallable'``, the Wanted arising from the
+   occurrence of ``uncallable`` is solved using the Given in the context.
+
+#. The following definition results in a custom type error message (as the
+   ``Unsatisfiable Msg`` constraint is reported rather than being generalised
+   over)::
+
+     rejected = uncallable  -- error
+
+#. The following definition is accepted::
+
+     unusual :: Unsatisfiable Msg => Char
+     unusual = 42  -- no error
+
+   Here the presence of ``Unsatisfiable Msg`` in the context means that the ``Num
+   Char`` constraint arising from the body of ``unusual`` is discharged
+   automatically.
```

Comment:
How would this example be desugared? I think something about desugaring is missing from the specification (it would handle this case, as well as `-fdefer-type-errors`)